pkg/executor/sortexec/parallel_sort_worker.go

```diff
@@ -0,0 +1,217 @@
+		}
+
+		// Memory usage of the chunk has been consumed at the producer side.
+		chk, ok := <-p.chunkChannel
+		if !ok {
+			return true
+		}
+
+		p.totalMemoryUsage += chk.MemoryUsage
+
+		sortedRows := p.sortChunkAndGetSortedRows(chk.Chk)
+		p.globalSortedRowsQueue.add(sortedRows)
+		p.injectFailPointForParallelSortWorker()
+	}
+}
+
+func (p *parallelSortWorker) sortChunkAndGetSortedRows(chk *chunk.Chunk) sortedRows {
+	rowNum := chk.NumRows()
+	p.rowBuffer = make(sortedRows, rowNum)
+	for i := 0; i < rowNum; i++ {
+		p.rowBuffer[i] = chk.GetRow(i)
+	}
+	slices.SortFunc(p.rowBuffer, p.keyColumnsLess)
+	return p.rowBuffer
+}
+
+func (p *parallelSortWorker) mergeSortGlobalRows() {
+	for {
+		err := p.checkError()
+		if err != nil {
+			return
+		}
+
+		sortedRowsLeft, sortedRowsRight := p.globalSortedRowsQueue.fetchTwoSortedRows()
+		if sortedRowsLeft == nil {
+			break
+		}
+
+		mergedSortedRows := p.mergeTwoSortedRows(sortedRowsLeft, sortedRowsRight)
+		p.globalSortedRowsQueue.add(mergedSortedRows)
+		p.injectFailPointForParallelSortWorker()
+	}
+}
+
+func (p *parallelSortWorker) mergeTwoSortedRows(sortedRowsLeft sortedRows, sortedRowsRight sortedRows) sortedRows {
+	sortedRowsLeftLen := len(sortedRowsLeft)
+	sortedRowsRightLen := len(sortedRowsRight)
+	mergedSortedRows := make(sortedRows, 0, sortedRowsLeftLen+sortedRowsRightLen)
+	cursorLeft := 0  // Point to sortedRowsLeft
+	cursorRight := 0 // Point to sortedRowsRight
+
+	// Merge
+	for cursorLeft < sortedRowsLeftLen && cursorRight < sortedRowsRightLen {
+		if p.timesOfRowCompare >= SignalCheckpointForSort {
+			// Trigger Consume for checking the NeedKill signal
+			p.memTracker.Consume(1)
+			p.timesOfRowCompare = 0
+		}
+
+		failpoint.Inject("SignalCheckpointForSort", func(val failpoint.Value) {
+			if val.(bool) {
+				p.timesOfRowCompare += 1024
+			}
+		})
+		p.timesOfRowCompare++
+
+		if p.lessRowFunc(sortedRowsLeft[cursorLeft], sortedRowsRight[cursorRight]) < 0 {
+			mergedSortedRows = append(mergedSortedRows, sortedRowsLeft[cursorLeft])
+			cursorLeft++
+		} else {
+			mergedSortedRows = append(mergedSortedRows, sortedRowsRight[cursorRight])
+			cursorRight++
+		}
+	}
+
+	// Append the remaining rows
+	mergedSortedRows = append(mergedSortedRows, sortedRowsLeft[cursorLeft:]...)
+	mergedSortedRows = append(mergedSortedRows, sortedRowsRight[cursorRight:]...)
+
+	return mergedSortedRows
+}
+
+func (p *parallelSortWorker) keyColumnsLess(i, j chunk.Row) int {
+	if p.timesOfRowCompare >= SignalCheckpointForSort {
+		// Trigger Consume for checking the NeedKill signal
+		p.memTracker.Consume(1)
+		p.timesOfRowCompare = 0
+	}
+
+	failpoint.Inject("SignalCheckpointForSort", func(val failpoint.Value) {
+		if val.(bool) {
+			p.timesOfRowCompare += 1024
+		}
+	})
+	p.timesOfRowCompare++
+
+	return p.lessRowFunc(i, j)
+}
+
+func (p *parallelSortWorker) run() {
+	defer func() {
+		if r := recover(); r != nil {
+			processPanicAndLog(p.processError, r)
+			p.tryToCloseChunkChannel()
```

Comment:
WHY sortWorker need to call this?

Reply:
> WHY sortWorker need to call this?

In case all workers panic, no one will consume this chunkChannel and data fetcher will be blocked in the channel when the channel is full. So, we need to close this channel when worker panic.

Reply:
Closing this channel will make data fetcher panic

Reply:
> Closing this channel will make data fetcher panic

Yes, and it's ok, data fetcher will recover this panic.